// mongoose/lib/schema/index.js

/*!
 * Module exports.
 */

'use strict';

exports.Array = require('mongoose/lib/schema/array');
exports.BigInt = require('mongoose/lib/schema/bigint');
exports.Boolean = require('mongoose/lib/schema/boolean');
exports.Buffer = require('mongoose/lib/schema/buffer');
exports.Date = require('mongoose/lib/schema/date');
exports.Decimal128 = exports.Decimal = require('mongoose/lib/schema/decimal128');
exports.DocumentArray = require('mongoose/lib/schema/documentArray');
exports.Map = require('mongoose/lib/schema/map');
exports.Mixed = require('mongoose/lib/schema/mixed');
exports.Number = require('mongoose/lib/schema/number');
exports.ObjectId = require('mongoose/lib/schema/objectId');
exports.String = require('mongoose/lib/schema/string');
exports.Subdocument = require('mongoose/lib/schema/subdocument');
exports.UUID = require('mongoose/lib/schema/uuid');

// alias

exports.Oid = exports.ObjectId;
exports.Object = exports.Mixed;
exports.Bool = exports.Boolean;
exports.ObjectID = exports.ObjectId;
